benchmarks/struct_indices/spider/evaluate.py

```diff
@@ -0,0 +1,291 @@
+import argparse
+import ast
+import json
+import logging
+import os
+from typing import Dict, List
+
+from langchain import OpenAI
+from langchain.chat_models import ChatOpenAI
+from langchain.schema import HumanMessage
+from spider_utils import create_indexes, load_examples
+from tqdm import tqdm
+
+from gpt_index.indices.struct_store.sql import GPTSQLStructStoreIndex
+
+logging.getLogger("root").setLevel(logging.WARNING)
+
+answer_template = (
+    "Given an input question, first create a syntactically correct SQL "
+    "query to run, then look at the results of the query and return the answer. "
+    "Use the following format:\n"
+    "Question: Question here\n"
+    "SQLQuery: SQL Query to run\n"
+    "SQLResult: Result of the SQLQuery\n"
+    "Answer: Final answer here\n"
+    "Question: {question}\n"
+    "SQLQuery: {sql_query}\n"
+    "SQLResult: {sql_result}"
+    "Answer: "
+)
+
+match_template = """Given a question, a reference answer and a hypothesis answer, determine if the hypothesis answer is correct. Use the following format:
+
+Question: Question here
+ReferenceAnswer: Reference answer here
+HypothesisAnswer: Hypothesis answer here
+HypothesisAnswerCorrect: true or false
+
+Question: {question}
+ReferenceAnswer: {reference_answer}
+HypothesisAnswer: {hypothesis_answer}
+HypothesisAnswerCorrect: """
+
+
+def _answer(llm: ChatOpenAI, question: str, sql_query: str, sql_result: str) -> str:
```

Comment:
very nit: from below, seems like llm could be OpenAI or ChatOpenAI. If so, typing is wrong.

Reply:
True. I guess when I wrote it they haven't added gpt-4 support for completion API.

Reply:
Oh wait, it looks like OpenAI still doesn't support gpt-4 or gpt-3.5-turbo from their completion endpoint. So the type has to be either `langchain.chat_models.ChatOpenAI` or `langchain.llms.OpenAIChat`. I guess maybe it is simpler to just use one type here since these two are the same for the purpose of evaluation and are backed by the same endpoint.